File: Binary Tree Postorder Traversal.java
// leetcode 145 problem
// 求二叉树的后序遍历序列

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

// Recursive solution
class Solution {
    public List<Integer> postorderTraversal(TreeNode root) {
		List<Integer> pre = new LinkedList<Integer>();
		preHelper(root, pre);
		return pre;
	}
    
	public void preHelper(TreeNode root, List<Integer> pre) {
		if (root == null) return;
		preHelper(root.left, pre);
		preHelper(root.right, pre);
        pre.add(root.val);
	}
}

//Iterative solution
public List<Integer> postorderTraversal(TreeNode root) {
    LinkedList<Integer> result = new LinkedList<>();
    Deque<TreeNode> stack = new ArrayDeque<>();
    TreeNode p = root;
    while(!stack.isEmpty() || p != null) {
        if(p != null) {
            stack.push(p);
            result.addFirst(p.val);  // Reverse the process of preorder
            p = p.right;             // Reverse the process of preorder
        } else {
            TreeNode node = stack.pop();
            p = node.left;           // Reverse the process of preorder
        }
    }
    return result;
}

File: Binary Tree Postorder Traversal.cpp
// leetcode 145 problem
// 二叉树后序遍历序列（迭代解法）

// 前序 根 左 右 -> 根 右 左 reverse-> 左 右 根 
class Solution {
public:
    vector<int> postorderTraversal(TreeNode *root) {
       vector<int> res;
       if (!root) return res;
       stack<TreeNode *> st;
       st.push(root);
       while (st.size()) {
           TreeNode *temp = st.top();
           st.pop();
           res.push_back(temp->val);
           if (temp->left)
               st.push(temp->left);
           if (temp->right)
               st.push(temp->right);
       }
       reverse(res.begin(),res.end());
       return res;
   }
};